hpat/datatypes/hpat_pandas_series_functions.py

```diff
@@ -138,6 +173,182 @@ def hpat_pandas_series_iloc_impl(self):
     return hpat_pandas_series_iloc_impl
 
 
+@overload_method(SeriesType, 'nsmallest')
+def hpat_pandas_series_nsmallest(self, n=5, keep='first'):
+    """
+    Pandas Series method :meth:`pandas.Series.nsmallest` implementation.
+
+    .. only:: developer
+       Test: python -m hpat.runtests -k hpat.tests.test_series.TestSeries.test_series_nsmallest*
+
+    n: :obj:`int`, default 5
+        Return this many ascending sorted values.
+    keep: :obj:`str`, default 'first'
+        When there are duplicate values that cannot all fit in a Series of n elements:
+        first : return the first n occurrences in order of appearance.
+        last : return the last n occurrences in reverse order of appearance.
+        all : keep all occurrences. This can result in a Series of size larger than n.
+        *unsupported*
+
+    Returns
+    -------
+    :obj:`series`
+         returns :obj:`series`
+    """
+
+    _func_name = 'Method nsmallest().'
+
+    ty_checker = TypingChecker(_func_name)
+    ty_checker.check_series(self)
+    ty_checker.check_integer(n, 'n')
+    ty_checker.check_string(keep, 'keep')
+
+    dtype = self.data.dtype
+    if isinstance(self.index, types.NoneType):
+        inner_list_type = types.ListType(types.intp)
+        def hpat_pandas_series_nsmallest_impl(self, n=5, keep='first'):
+            if keep != 'first':
+                raise ValueError("Method nsmallest(). Unsupported parameter. "
+                                 "Given 'keep' != 'first'")
+
+            nsmallest = hpat.hiframes.api.nlargest(self._data, n, False,
+                                                   hpat.hiframes.series_kernels.lt_f)
+            indices = Dict.empty(dtype, inner_list_type)
+            for idx, item in enumerate(self._data):
+                if item not in indices:
+                    indices[item] = List.empty_list(types.intp)
+                indices[item].append(idx)
+
+            max_length = len(nsmallest)
+            local_index = []
+            for item in numpy.unique(nsmallest):
+                for index in indices[item]:
+                    if len(local_index) == max_length:
+                        break
+                    local_index.append(index)
```

Comment:
I suspect this algorithm might be replaced by `numpy.argsort`

Reply:
I replaced the algorithm with `numpy.argsort` for `nsmallest`. I'm not able to apply that for `nlargest` due to descending ordering of the result (`[::-1]` works incorrectly).

Reply:
@shssf Why to use argsort here? It will sort the whole array whereas we need just first n elements?

Reply:
Now I'm not able to explain it, but looks like implementation with using of `argsort` is up to 3 times faster on laptop on 4 threads.

name | type | size | n | median | min | max
-- | -- | -- | -- | -- | -- | --
series_nsmallest | JIT | 10002024 | 16 | 1,928377 | 1,914718 | 1,956772
  |   |   | 64 | 1,878858 | 1,866676 | 1,910763
  |   |   | 512 | 1,869665 | 1,853625 | 2,024022
  |   |   | 1024 | 1,877665 | 1,844636 | 1,878451
series_nsmallest   (with arsort) | JIT | 10002024 | 16 | 0,722629 | 0,717613 | 0,76169
  |   |   | 64 | 0,725377 | 0,71963 | 0,765624
  |   |   | 512 | 0,738325 | 0,725678 | 0,743378
  |   |   | 1024 | 0,745736 | 0,730581 | 0,759536